yjit/src/core.rs

```diff
@@ -286,63 +287,85 @@ pub enum TypeDiff {
 #[derive(Copy, Clone, Eq, Hash, PartialEq, Debug)]
-pub enum LocalIndex {
-    Local0,
-    Local1,
-    Local2,
-    Local3,
-    Local4,
-    Local5,
-    Local6,
-    Local7,
-}
+pub struct TempMapping(u8);
 
-impl From<LocalIndex> for u8 {
-    fn from(idx: LocalIndex) -> Self {
-        match idx {
-            LocalIndex::Local0 => 0,
-            LocalIndex::Local1 => 1,
-            LocalIndex::Local2 => 2,
-            LocalIndex::Local3 => 3,
-            LocalIndex::Local4 => 4,
-            LocalIndex::Local5 => 5,
-            LocalIndex::Local6 => 6,
-            LocalIndex::Local7 => 7,
-        }
+impl TempMapping {
+    pub fn map_to_stack(t: Type) -> TempMapping
+    {
+        let kind_bits = unsafe { transmute::<TempMappingKind, u8>(TempMappingKind::MapToStack) };
+        let type_bits = unsafe { transmute::<Type, u8>(t) };
+        assert!(type_bits < 32);
+        let bits = (kind_bits << 6) | (type_bits & 31);
+        TempMapping(bits)
+    }
+
+    pub fn map_to_self() -> TempMapping
+    {
+        let kind_bits = unsafe { transmute::<TempMappingKind, u8>(TempMappingKind::MapToSelf) };
+        let bits = kind_bits << 6;
+        TempMapping(bits)
+    }
+
+    pub fn map_to_local(local_idx: u8) -> TempMapping
+    {
+        let kind_bits = unsafe { transmute::<TempMappingKind, u8>(TempMappingKind::MapToLocal) };
+        assert!(local_idx < 8);
+        let bits = (kind_bits << 6) | (local_idx & 7);
+        TempMapping(bits)
     }
-}
 
-impl From<u8> for LocalIndex {
-    fn from(idx: u8) -> Self {
-        match idx {
-            0 => LocalIndex::Local0,
-            1 => LocalIndex::Local1,
-            2 => LocalIndex::Local2,
-            3 => LocalIndex::Local3,
-            4 => LocalIndex::Local4,
-            5 => LocalIndex::Local5,
-            6 => LocalIndex::Local6,
-            7 => LocalIndex::Local7,
-            _ => unreachable!("{idx} was larger than {MAX_LOCAL_TYPES}"),
+    pub fn without_type(&self) -> TempMapping
+    {
+        if self.get_kind() != TempMappingKind::MapToStack {
+            return *self;
         }
+
+        TempMapping::map_to_stack(Type::Unknown)
+    }
+
+    pub fn get_kind(&self) -> TempMappingKind
+    {
+        // Take the two highest bits
+        let TempMapping(bits) = self;
+        let kind_bits = bits >> 6;
+        assert!(kind_bits <= 2);
+        unsafe { transmute::<u8, TempMappingKind>(kind_bits) }
+    }
+
+    pub fn get_type(&self) -> Type
+    {
+        assert!(self.get_kind() == TempMappingKind::MapToStack);
+
+        // Take the 5 lowest bits
+        let TempMapping(bits) = self;
+        let type_bits = bits & 31;
+        unsafe { transmute::<u8, Type>(type_bits) }
+    }
+
+    pub fn get_local_idx(&self) -> u8
+    {
+        assert!(self.get_kind() == TempMappingKind::MapToLocal);
+
+        // Take the 3 lowest bits
+        let TempMapping(bits) = self;
+        bits & 7
```

Comment:
Maybe nice to use binary notation in this context?
```suggestion
        bits & 0b111
```